[TZone] Update TZone debugging code

msaboff opened the pull request:
#### 7098ad895d64acbb6fae3637fbc72cc53f777300
<pre>
[TZone] Update TZone debugging code
<a href="https://bugs.webkit.org/show_bug.cgi?id=281641">https://bugs.webkit.org/show_bug.cgi?id=281641</a>
<a href="https://rdar.apple.com/138083630">rdar://138083630</a>

Reviewed by Yijia Huang.

Added various debugging improvements to the TZone manager and allocation path for unannotated subclasses.
The changes for &quot;verbose&quot; compilation mode include:
 * Adding bucket counts for each size class.
 * Simplifying the data needed for statistics.
 * Using a separate verbose flag for the dump of bucket selection.
 * Moving all debug code to a .cpp file to speed up compilation when changing debug flags.

For the unannotated subclass allocation path, split the debug info into two lines so that the info isn’t truncated in the OS log.
Also fixed the spelling of the method dumpRegisteredTypes.

* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::api::TZoneHeapBase::provideHeap):
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::dumpRegisteredTypesAtExit):
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::dumpRegisteredTypes):
(bmalloc::api::TZoneHeapManager::tzoneBucketForKey):
(bmalloc::api::TZoneHeapManager::populateBucketsForSizeClass):
(bmalloc::api::TZoneHeapManager::heapRefForTZoneType):
(bmalloc::api::dumpRegisterdTypesAtExit): Deleted.
(bmalloc::api::TZoneHeapManager::dumpRegisterdTypes): Deleted.
* Source/bmalloc/bmalloc/TZoneHeapManager.h:
(bmalloc::api::TZoneHeapManager::tzoneBucketForKey): Deleted.

Canonical link: <a href="https://commits.webkit.org/285355@main">https://commits.webkit.org/285355@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3603fd44b47a6442c50693b72d617bb9b065013

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72234 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51655 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25022 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76402 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23443 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59459 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23265 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56937 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15443 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75301 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46796 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37376 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19678 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21793 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65363 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65358 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78079 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71488 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19197 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65403 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16522 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64666 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15986 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12887 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6532 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93269 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47453 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20530 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/verbose-failure-dont-graph-dump-availability-already-freed.js.default, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48522 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49810 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48265 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
```diff
@@ -141,7 +158,8 @@ void TZoneHeapManager::init()
     m_state = TZoneHeapManager::Seeded;
 
-    atexit(dumpRegisterdTypesAtExit);
+    if (verbose)
```
I think we can use `constexpr` here?

msaboff replied:
Actually, we can't.  With constexpr, the compiler will eliminate the call to atexit(dumpRegisterdTypesAtExit); and then complain that dumpRegisterdTypesAtExit() is unused.

hyjorc1 reviewed (approved):
r=me

msaboff committed:
[TZone] Update TZone debugging code
https://bugs.webkit.org/show_bug.cgi?id=281641
rdar://138083630

Reviewed by Yijia Huang.

Added various debugging improvements to the TZone manager and allocation path for unannotated subclasses.
The changes for "verbose" compilation mode include:
 * Adding bucket counts for each size class.
 * Simplifying the data needed for statistics.
 * Using a separate verbose flag for the dump of bucket selection.
 * Moving all debug code to a .cpp file to speed up compilation when changing debug flags.

For the unannotated subclass allocation path, split the debug info into two lines so that the info isn’t truncated in the OS log.
Also fixed the spelling of the method dumpRegisteredTypes.

* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::api::TZoneHeapBase::provideHeap):
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::dumpRegisteredTypesAtExit):
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::dumpRegisteredTypes):
(bmalloc::api::TZoneHeapManager::tzoneBucketForKey):
(bmalloc::api::TZoneHeapManager::populateBucketsForSizeClass):
(bmalloc::api::TZoneHeapManager::heapRefForTZoneType):
(bmalloc::api::dumpRegisterdTypesAtExit): Deleted.
(bmalloc::api::TZoneHeapManager::dumpRegisterdTypes): Deleted.
* Source/bmalloc/bmalloc/TZoneHeapManager.h:
(bmalloc::api::TZoneHeapManager::tzoneBucketForKey): Deleted.

Canonical link: https://commits.webkit.org/285355@main
